rumqttd/src/segments/mod.rs

```diff
@@ -272,10 +274,17 @@ where
             }
             SegmentPosition::Done(absolute_offset) => {
                 // debug!("start: {:?}, end: ({}, {}) done", orig_cursor, cursor.0, absolute_offset);
-                Ok(Position::Done {
-                    start,
-                    end: (cursor.0, absolute_offset),
-                })
+                if self.active_segment().size() < self.max_segment_size as u64 {
+                    Ok(Position::Done {
+                        start,
+                        end: (cursor.0, absolute_offset),
+                    })
+                } else {
+                    Ok(Position::Done {
+                        start,
+                        end: (cursor.0 + 1, absolute_offset),
+                    })
+                }
```

Comment:
if `active_segment().size()` >= `self.max_segment_size` on the next [append to commitlog](https://github.com/bytebeamio/rumqtt/blob/3f388b00f2e4eb8d7e710673bbc0dcb137244bbc/rumqttd/src/segments/mod.rs#L161) we will call `apply_retention` and new segment would be created.